stdlib/public/core/Result.swift

```diff
@@ -0,0 +1,80 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+@_frozen
+public enum Result<Value, Error> {
+    case success(Value)
+    case failure(Error)
+    
+    public var value: Value? {
+        switch self {
+        case let .success(value): return value
+        case .failure: return nil
+        }
+    }
+    
+    public var error: Error? {
+        switch self {
+        case let .failure(error): return error
+        case .success: return nil
+        }
+    }
+    
+    public var isSuccess: Bool {
+        switch self {
+        case .success: return true
+        case .failure: return false
+        }
+    }
+    
+}
+
+extension Result where Error: Swift.Error {
+    public func unwrap() throws -> Value {
+        switch self {
+        case let .success(value): return value
+        case let .failure(error): throw error
+        }
+    }
+}
+
+extension Result where Error == Swift.Error {
+    @_transparent
+    public init(_ throwing: () throws -> Value) {
+        do {
+            let value = try throwing()
+            self = .success(value)
+        } catch {
+            self = .failure(error)
+        }
+    }
+    
+    public func unwrap() throws -> Value {
+        switch self {
+        case let .success(value): return value
+        case let .failure(error): throw error
+        }
+    }
+}
+
+extension Result: Equatable where Value: Equatable, Error: Equatable {
+    public static func ==(lhs: Result<Value, Error>, rhs: Result<Value, Error>) -> Bool {
+        return lhs.value == rhs.value && lhs.error == rhs.error
+    }
+}
+
+extension Result: Hashable where Value: Hashable, Error: Hashable {
+    public func hash(into hasher: inout Hasher) {
+        hasher.combine(value)
```

Comment:
Similar to my above comment, would it not be better to do a switch on `self` here? Of course one of the cases would need to add some extra bit to the hasher in order to ensure that the hashes are different when the `Error` type and the `Value` are the same and the values are equal...

Reply:
I'm not sure. If it's more efficient, then I'll make the change, but can I annotate the properties to regain the efficiency?